### ISSUE TITLE
Java SDK v2 support is broken

### ISSUE THREAD

**Issue by nyh** [scylladb/alternator-load-balancing#15]
A user reported failure use our support for the AWS Java SDK v2 on recent versions of the SDK. SDK versions 2.20.7 and 2.25.29 were tested and failed, with our Demo2.java example or the following simple example:
```java
public class ScyllaDbTest {
  public static void main(String[] args) {
    URI url = URI.create("http://localhost:8000/");
    DynamoDbClient ddb = AlternatorClient.builder(url.getScheme(), url.getHost(), url.getPort())
                                         .credentialsProvider(DefaultCredentialsProvider.create())
                                         .build();

    for (int i=0; i<10; i++) {
      DescribeEndpointsRequest request = DescribeEndpointsRequest.builder().build();
      DescribeEndpointsResponse response = ddb.describeEndpoints(request);
      System.out.println(response);
    }
  }
}
```
Java 17 was used.

The error is:
```
[main] WARN com.scylladb.alternator.AlternatorApacheClient - SSL Certificate verification is disabled. This is not a safe setting and should only be used for testing.
Apr 11, 2024 1:57:04 PM com.scylladb.alternator.AlternatorLiveNodes run
INFO: AlternatorLiveNodes thread starting
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "software.amazon.awssdk.metrics.MetricCollector.reportMetric(software.amazon.awssdk.metrics.SdkMetric, Object)" because "metricCollector" is null
	at software.amazon.awssdk.http.apache.internal.conn.ClientConnectionRequestFactory$InstrumentedConnectionRequest.get(ClientConnectionRequestFactory.java:73)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:190)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56)
	at software.amazon.awssdk.http.apache.internal.impl.ApacheSdkHttpClient.execute(ApacheSdkHttpClient.java:72)
	at com.scylladb.alternator.AlternatorApacheClient.execute(AlternatorApacheClient.java:179)
	at com.scylladb.alternator.AlternatorApacheClient.access$200(AlternatorApacheClient.java:90)
	at com.scylladb.alternator.AlternatorApacheClient$1.call(AlternatorApacheClient.java:159)
	at com.scylladb.alternator.AlternatorApacheClient$1.call(AlternatorApacheClient.java:156)
	at software.amazon.awssdk.core.internal.util.MetricUtils.measureDurationUnsafe(MetricUtils.java:63)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.executeHttpRequest(MakeHttpRequestStage.java:77)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:56)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:39)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:73)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:78)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:40)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:50)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:81)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:56)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:48)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:31)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
	at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:193)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:103)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:171)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:82)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:179)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:76)
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:56)
	at software.amazon.awssdk.services.dynamodb.DefaultDynamoDbClient.describeEndpoints(DefaultDynamoDbClient.java:1743)
```

**Comment by nyh** [comment]
I can reproduce this issue: Out of the box, our Demo2 demo still works (`mvn exec:java -Dexec.mainClass=com.scylladb.alternator.test.Demo2 -Dexec.classpathScope=test`, see the README.md), but this is because because our pom.xml fixed the SDK version at 2.15.46 - which works. But if the pom is updated to ask for SDK version 2.25.29 (as in the above report), I get a similar NullPointerException to what was reported above.

**Comment by nyh** [comment]
The cause of this issue is frustrating. The way we (Piotr Sarna) did the Java SDK 2 support was to basically copy and fork an internal SDK class ApacheHttpClient (our fork is AlternatorApacheClient). Not very suprisingly, the SDK's implementation of this class changed - they now have code that sets a new internal `THREAD_LOCAL_REQUEST_METRIC_COLLECTOR` for every request, and without it, the deeper level of the implementation (which we did NOT fork) failed, on that NullPointerException.

I could fix this problem by making a copy of the **new** ApacheHttpClient instead of the old one, but a better solution would not involve copying SDK code at all. I want to consider this approach version, and only if I fail, go back to copying code wholesale :-(

**Comment by nyh** [comment]
I found a much, much simpler solution: The Java SDK v2 has a new feature, DynamoDbEndpointProvider (https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/dynamodb/endpoints/DynamoDbEndpointProvider.html) - which seems to be called separately for every request and can return a different endpoint every time.  A quick test showed this actually works! I need to do more testing to verifies it really does work correctly, but it seems very promising - it's simplifies everything, doesn't duplicate and SDK code, and even simpler than the other new SDK v2 feature that I wanted to use - ExecutionInterceptor - but is much more general (and I'll only need to consider it if EndpointProvider doesn't work).

**Comment by nyh** [comment]
By the way, I checked, and my fix uses a new API that was added in Java SDK 2.20, released in February 2023. This is also the version that broke the old, complicated, code. In theory we could have had code that supported both the old and the new versions of the SDK (both of them SDK v2, mind you), but I don't want to do this unless we absolutely must. I'm planning to send a PR that will only work on SDK 2.20 and above and drop the ugly support for the newer versions. I assume that customers can use a modern version of the SDK and won't stick to a 2 year old one. I hope I'm assuming right - if not we can always revert my patch...
